core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala

```diff
@@ -5,14 +5,152 @@ package pureconfig.error
 
 import com.typesafe.config.ConfigValue
 
-sealed abstract class ConfigReaderFailure
-final case object CannotConvertNull extends ConfigReaderFailure
-final case class CannotConvert(value: String, toTyp: String, because: String) extends ConfigReaderFailure
-final case class CollidingKeys(key: String, existingValue: String) extends ConfigReaderFailure
-final case class KeyNotFound(key: String) extends ConfigReaderFailure
-final case class UnknownKey(key: String) extends ConfigReaderFailure
-final case class WrongType(foundTyp: String, expectedTyp: String) extends ConfigReaderFailure
-final case class WrongTypeForKey(typ: String, expectedTyp: String, key: String) extends ConfigReaderFailure
-final case class ThrowableFailure(throwable: Throwable) extends ConfigReaderFailure
-final case class EmptyStringFound(typ: String) extends ConfigReaderFailure
-final case class NoValidCoproductChoiceFound(value: ConfigValue) extends ConfigReaderFailure
\ No newline at end of file
+/**
+ * The physical location of a ConfigValue, represented by a file name and a line
+ * number
+ *
+ * @param filename the complete pathname to the file that contains a given
+ *                 ConfigValue
+ * @param lineNumber the line number (starting at 0), where the given
+ *                   ConfigValue definition starts
+ */
+case class ConfigValueLocation(filename: String, lineNumber: Int) {
+  def description: String = s"(file:$filename: $lineNumber)"
+}
+
+object ConfigValueLocation {
+  /**
+   * Helper method to create an optional ConfigValueLocation from a ConfigValue.
+   * Since it might not be possible to derive a ConfigValueLocation from a
+   * ConfigValue, this method returns Option.
+   *
+   * @param cv the ConfigValue to derive the location from
+   *
+   * @return a Some with the location of the ConfigValue, or None if it is not
+   *         possible to derive a location. It is not possible to derive
+   *         locations from ConfigValues that are not in files or for
+   *         ConfigValues that are null.
+   */
+  def apply(cv: ConfigValue): Option[ConfigValueLocation] =
+    Option(cv).flatMap { v =>
+      val origin = v.origin()
+      if (origin.filename != null && origin.lineNumber != -1)
+        // I believe lineNumber starts at 1 in ConfigOrigin
+        Some(ConfigValueLocation(origin.filename, origin.lineNumber - 1))
+      else
+        None
+    }
+}
+
+/**
+ * A representation of a failure that might be raised from reading a
+ * ConfigValue. The failure contains an optional location of the ConfigValue
+ * that raised the error.
+ */
+sealed abstract class ConfigReaderFailure {
+  /**
+   * The optional location of the ConfigReaderFailure
+   */
+  def location: Option[ConfigValueLocation]
+}
+
+/**
+ * A failure representing the inability of converting a null value. Since a null
```

Comment:
Nitpick: Here and below in a place or two, I'd suggest the phrase "the inability to convert" instead of "the inability of converting".